waspc/examples/todoApp/src/shared/webSocket.ts

```diff
@@ -0,0 +1,13 @@
+import { WaspSocketData } from '@wasp/universal/types'
```

Comment:
We could also do full-stack types here:
```ts
// Server is coming from Socket.IO
import { webSocketFn, Server } from '../../../server/src/ext-src/websocket'

type WebSocketFn = typeof webSocketFn

type Events = Parameters<WebSocketFn>[0] extends Server<infer F, infer S> ? [F, S] : never

const MainPage = ({ user }) => {
  const { socket, isConnected } = useSocket<Events[1], Events[0]>();
```

This is very rough but it works:
1. uses the type of the user defined webSocket function
2. Infers the two generic arguments the user passed when defining the fn

Reply:
That way we avoid having the user import the same types in the client and the server, they can just define it on the server

Reply:
Interesting. So would Wasp be the one doing the above, or the user? I am unsure what we would expose or how the user would still parameterize `useSocket` in a clean way based on the example?

Also, does this mean we would be including `Server` (and thus the non-client `'socket.io'`) on the client, too?

Reply:
Oh fancy I don't understand it fully but it sounds great :D.

Reply:
Question -> for every usage of `useSocket`, they will have to provide these types, right? Isn't that a bit cumbersome?

What if we instead let them provide those types to Wasp, and then Wasp will make sure to pre-type the `useSocket`, so they don't have to specify these types on it?

@infomiho ?